### PR TITLE
feat(ton): TON localnet integration

### PR DESCRIPTION
**Pull request by swift1337**
# Description

This PR adds containerized TON localnet node with all wiring for local development and E2E testing.

Note that the build is `x86` only.

- [x] Dockerize TON localnet
- [x] Add a sidecar with various features
  - [x] Get faucet
  - [x] Get lite client config
  - [x] Get `/status` healthcheck
- [x] Setup boilerplate for E2E TON testing suite with 1 hello world test for TON
- [x] `lite-client.json`: swap `localhost` (encoded as [int](https://www.browserling.com/tools/dec-to-ip) `2130706433`) with `ton`'s container IP 🪖 
- [x] Ability to wait for TON node bootstrapping
- [x] CI: add TON tests as `make start-ton-test` triggered with the label **TON_TESTS**
- [x] Move `ton` docker image to a separate repo

## Utils
- [x] Act on behalf of the faucet
- [x] Deploy arbitrary wallets
- [x] Deploy gateway
- [x] Fund arbitrary wallets

Based on https://github.com/neodix42/MyLocalTon

Closes: https://github.com/zeta-chain/node/issues/2720

## Summary by CodeRabbit

- **New Features**
	- Introduced support for TON blockchain with a dedicated testing framework.
	- Added a new service for TON in the local network setup.
	- Enhanced configuration options for managing TON accounts and RPC interactions.
	- Implemented end-to-end tests for TON deposits.

- **Bug Fixes**
	- Improved error handling and logging for TON-related operations.

- **Documentation**
	- Updated changelog to reflect new features and testing capabilities for TON.

- **Tests**
	- Added new test cases for TON functionality and wallet management.

### CODE REVIEW

File: contrib/localnet/ton/Dockerfile
Comment (gartnera) [inline contrib/localnet/ton/Dockerfile:1]: Would be nice to do this docker build somewhere else like solana: https://github.com/zeta-chain/solana-docker

Comment (swift1337) [reply contrib/localnet/ton/Dockerfile:1]: Yes, I'll definitely move this to a separate repo after everything works 👍

File: zetaclient/chains/ton/utils.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package ton
```
Comment (lumtis) [inline zetaclient/chains/ton/utils.go:1]: Can we avoid using utils file? Maybe rename to config

**Review by lumtis — APPROVED** [review]
Small comments look good to me

File: contrib/localnet/ton/readme.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+- Please note that it might take **several minutes** to bootstrap the network.
+
+## Sidecar
+
+### Getting faucet wallet
+
+```shell
+curl -s http://ton:8000/faucet.json | jq
+{
+  "initialBalance": 1000001000000000,
```
[the remark this answers is not in the document]
Comment (swift1337) [reply contrib/localnet/ton/readme.md:19]: It comes from https://github.com/neodix42/MyLocalTon

File: contrib/localnet/docker-compose.yml [hunk 3]
```diff
@@ -241,6 +241,22 @@ services:
...
     entrypoint: [ "/usr/bin/start-solana.sh" ]
 
+  ton:
+    image: ghcr.io/zeta-chain/ton-docker:latest
+    container_name: ton
+    hostname: ton
+    profiles:
+      - ton
+      - all
+    ports:
```
Comment (lumtis) [inline contrib/localnet/docker-compose.yml:251]:
```suggestion
    platform: linux/amd64
    ports:
```
This fix an error on MacOS
```
no matching manifest for linux/arm64/v8 in the manifest list entries
```

Comment (swift1337) [reply contrib/localnet/docker-compose.yml:251]: It's going to be fixed with https://github.com/zeta-chain/ton-docker/pull/2

File: cmd/zetae2e/local/ton.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+	deployerRunner *runner.E2ERunner,
+	verbose bool,
+	testNames ...string,
+) func() error {
+	return func() (err error) {
+		tonRunner, err := initTestRunner(
+			"ton",
+			conf,
+			deployerRunner,
+			conf.AdditionalAccounts.UserSolana,
```
[the remark this answers is not in the document]
Comment (swift1337) [reply cmd/zetae2e/local/ton.go]: It's a typo. Thanks for spotting this

File: contrib/localnet/docker-compose.yml [hunk 5]
```diff
@@ -31,7 +31,7 @@ services:
     networks:
       mynetwork:
         ipv4_address: 172.20.0.200
-    entrypoint: ["zetacored", "rosetta", "--tendermint", "zetacore0:26657", "--grpc", "zetacore0:9090", "--network", "athens_101-1", "--blockchain",  "zetacore"]
+    entrypoint: [ "zetacored", "rosetta", "--tendermint", "zetacore0:26657", "--grpc", "zetacore0:9090", "--network", "athens_101-1", "--blockchain",  "zetacore" ]
```
Comment (gartnera) [inline contrib/localnet/docker-compose.yml:34]: What are these formatting changes?

**Review by gartnera — APPROVED** [review]
CI stamp

File: contrib/localnet/docker-compose.yml (hunk 5 above)
Comment (swift1337) [reply contrib/localnet/docker-compose.yml]: https://github.com/zeta-chain/node/pull/2833/commits/ff7a62ebe8d4e0c9cf4af793a8e3a062afb8b696

File: .github/workflows/e2e.yml
[the remark this answers is not in the document]
Comment (swift1337) [reply .github/workflows/e2e.yml:1]: https://github.com/zeta-chain/node/pull/2833/commits/ee7ed8b1a9e626b97497f0c7c966e119dc0ebbf6

**Pull request merged**